src/java/org/apache/cassandra/cql3/statements/ListSuperUsersStatement.java

```diff
@@ -0,0 +1,108 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.cql3.statements;
+
+import java.util.List;
+import java.util.Set;
+
+import com.google.common.collect.ImmutableList;
+import org.apache.commons.lang3.builder.ToStringBuilder;
+import org.apache.commons.lang3.builder.ToStringStyle;
+
+import org.apache.cassandra.audit.AuditLogContext;
+import org.apache.cassandra.audit.AuditLogEntryType;
+import org.apache.cassandra.auth.AuthKeyspace;
+import org.apache.cassandra.auth.Permission;
+import org.apache.cassandra.auth.RoleResource;
+import org.apache.cassandra.auth.Roles;
+import org.apache.cassandra.config.DatabaseDescriptor;
+import org.apache.cassandra.cql3.ColumnIdentifier;
+import org.apache.cassandra.cql3.ColumnSpecification;
+import org.apache.cassandra.cql3.ResultSet;
+import org.apache.cassandra.db.marshal.UTF8Type;
+import org.apache.cassandra.exceptions.InvalidRequestException;
+import org.apache.cassandra.exceptions.RequestExecutionException;
+import org.apache.cassandra.exceptions.RequestValidationException;
+import org.apache.cassandra.exceptions.UnauthorizedException;
+import org.apache.cassandra.schema.SchemaConstants;
+import org.apache.cassandra.service.ClientState;
+import org.apache.cassandra.transport.messages.ResultMessage;
+
+/**
+ * LIST SUPERUSERS cql command returns list of roles with superuser privileges
+ * This includes superusers and all roles who have superuser role granted in the roles hierarchy
+ */
+public class ListSuperUsersStatement extends AuthorizationStatement
+{
+    private static final List<ColumnSpecification> metadata =
+    ImmutableList.of(new ColumnSpecification(SchemaConstants.AUTH_KEYSPACE_NAME, AuthKeyspace.ROLES,
+                                             new ColumnIdentifier("role", true), UTF8Type.instance));
+
+    public ListSuperUsersStatement()
+    {
+        // nothing to do
+    }
+
+    public void validate(ClientState state) throws UnauthorizedException, InvalidRequestException
+    {
+        state.ensureNotAnonymous();
+    }
+
+    public void authorize(ClientState state) throws InvalidRequestException
+    {
+        // Allow listing superuser privileged users only if the caller has DESCRIBE permission on 'all roles'
+        if (!DatabaseDescriptor.getAuthorizer()
+                               .authorize(state.getUser(), RoleResource.root())
+                               .contains(Permission.DESCRIBE))
+        {
+            throw new UnauthorizedException("You are not authorized to view superuser details");
+        }
+    }
+
+    public ResultMessage execute(ClientState state) throws RequestValidationException, RequestExecutionException
+    {
+        return resultMessage(Roles.getAllSuperUsers());
+    }
+
+    private ResultMessage resultMessage(Set<RoleResource> roles)
```

Comment:
can't be the body of this method just merged with `execute` method? I just dont see the advantage of moving the actual logic into this method when it can be just all done in `execute`.

Something like this:

    public ResultMessage execute(ClientState state) throws RequestValidationException, RequestExecutionException
    {
        Set<RoleResource> allSuperusers = Roles.getAllRoles(Roles::hasSuperuserStatus);

        if (allSuperusers.isEmpty())
            return new ResultMessage.Void();

        ResultSet result = new ResultSet(new ResultSet.ResultMetadata(metadata));

        allSuperusers.stream()
                     .sorted(RoleResource::compareTo)
                     .forEach(role -> result.addColumnValue(UTF8Type.instance.decompose(role.getRoleName())));

        return new ResultMessage.Rows(result);
    }

Reply:
Following the same style as LIST ROLES command is doing. This separation usually helps when a command has different clauses supported and formatting differently for different clauses. In this case (as of now), it's quite simple without any clauses, so we can merge

Reply:
yes please merge, this is very straightforward logic

Reply:
Done